In [62]:
import pandas as pd
import numpy as np

### Задание 1. Загрузка данных
Изучить представленный набор данных на основе описания его столбцов, загрузить его и оставить 8 столбцов для дальнейшего изучения: surgery?, Age, rectal temperature, pulse, respiratory rate, temperature of extremities, pain, outcome.

### Задание 2. Первичное изучение данных
Проанализировать значения по столбцам, рассчитать базовые статистики, найти выбросы.

### Задание 3. Работа с пропусками
Рассчитать количество пропусков для всех выбранных столбцов. Принять и обосновать решение о методе заполнения пропусков по каждому столбцу на основе рассчитанных статистик и возможной взаимосвязи значений в них. Сформировать датафрейм, в котором пропуски будут отсутствовать.

In [63]:
# surgery?(0), Age(1), rectal temperature(3), pulse(4), respiratory rate(5), temperature of extremities(6), pain(10), outcome(22).

In [64]:
df_start = pd.read_csv('horse_data.csv', sep=',', header = None, na_values = '?')
df_start = df_start.iloc[:, [0, 1, 3, 4, 5, 6, 10, 22]]
df_start.columns = ['surgery?', 'age', 'rectal temperature', 'pulse', 'respiratory rate', 'temperature of extremities', 'pain', 'outcome']
df_start

surgery?  age  rectal temperature  pulse  respiratory rate  \
0         2.0    1                38.5   66.0              28.0   
1         1.0    1                39.2   88.0              20.0   
2         2.0    1                38.3   40.0              24.0   
3         1.0    9                39.1  164.0              84.0   
4         2.0    1                37.3  104.0              35.0   
..        ...  ...                 ...    ...               ...   
295       1.0    1                 NaN  120.0              70.0   
296       2.0    1                37.2   72.0              24.0   
297       1.0    1                37.5   72.0              30.0   
298       1.0    1                36.5  100.0              24.0   
299       1.0    1                37.2   40.0              20.0   

     temperature of extremities  pain  outcome  
0                           3.0   5.0      2.0  
1                           NaN   3.0      3.0  
2                           1.0   3.0      1.0  
3                           4.0   2.0      2.0  
4                           NaN   NaN      2.0  
..                          ...   ...      ...  
295                         4.0   2.0      3.0  
296                         3.0   4.0      3.0  
297                         4.0   4.0      2.0  
298                         3.0   3.0      1.0  
299                         NaN   NaN      3.0  

[300 rows x 8 columns]

 surgery?(0) - 1 или 2  
 Age(1) - 1 или 2  
 rectal temperature(3) - норма 37,8  
 pulse(4) - норма 20-40  
 respiratory rate(5) - норма 8-10  
 temperature of extremities(6) - от 1до 4  
 pain(10) - от 1 до 5  
 outcome(22) - 1, 2 или 3  

In [65]:
pd.concat([df_start.describe(), df_start.mode()])

surgery?         age  rectal temperature       pulse  \
count  299.000000  300.000000          240.000000  276.000000   
mean     1.397993    1.640000           38.167917   71.913043   
std      0.490305    2.173972            0.732289   28.630557   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    9.000000           40.800000  184.000000   
0        1.000000    1.000000           38.000000   48.000000   

       respiratory rate  temperature of extremities       pain     outcome  
count        242.000000                  244.000000  245.00000  299.000000  
mean          30.417355                    2.348361    2.95102    1.551839  
std           17.642231                    1.045054    1.30794    0.737187  
min            8.000000                    1.000000    1.00000    1.000000  
25%           18.500000                    1.000000    2.00000    1.000000  
50%           24.500000                    3.000000    3.00000    1.000000  
75%           36.000000                    3.000000    4.00000    2.000000  
max           96.000000                    4.000000    5.00000    3.000000  
0             20.000000                    3.000000    3.00000    1.000000

Рассчитав базовые статистики мы можем сделать следующие выводы:
колонки хирургия, возраст и результат имеют не более 1 пропуска, но все они принимают бинарное значение 1 или 2, а на основании этих данных невозможно корректно рассчитать остальные величины, так же для заполнения пропусков сложно использовать значения колонок температура конечностей и боль, т.к. это субъективные оценки. Отсюда можем предположить, что имеет смысл оставить только те, строки, где заполнено минимум 6 значений из 8. Таким образом мы убираем 12 строк, но практически не влияем на базовые статистики.


In [66]:
df_start['age'].value_counts()

1    276
9     24
Name: age, dtype: int64

In [97]:
df = df_start.dropna(thresh=6)

In [98]:
pd.concat([df.describe(), df.mode()])

surgery?         age  rectal temperature       pulse  \
count  277.000000  278.000000          238.000000  269.000000   
mean     1.404332    1.633094           38.168908   72.055762   
std      0.491651    2.163511            0.734588   28.815170   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    9.000000           40.800000  184.000000   
0        1.000000    1.000000           38.000000   48.000000   

       respiratory rate  temperature of extremities        pain     outcome  
count        240.000000                  236.000000  238.000000  277.000000  
mean          30.454167                    2.338983    2.928571    1.548736  
std           17.687607                    1.053629    1.295539    0.743552  
min            8.000000                    1.000000    1.000000    1.000000  
25%           19.500000                    1.000000    2.000000    1.000000  
50%           24.500000                    3.000000    3.000000    1.000000  
75%           36.000000                    3.000000    4.000000    2.000000  
max           96.000000                    4.000000    5.000000    3.000000  
0             20.000000                    3.000000    3.000000    1.000000

### Выбросы

surgery? - выбросов нет
Age - мы видим, что все значения принимаю 1 или 9, и нет ни одного значения 2, отсюда можем предположить, что 9 это ошибка и их необходимо заменить на 2

rectal temperature - принимает значения от 35,4 до 40,8 при норме 37,8. 
pulse - от 30 до 184 пр норме 20-40
respiratory rate  - от 8 до 96 при норме 8-10
Зная, что речь идет не о здоровых лошадях, то можно предположить, что все эти значения возможны и не являются выбросами

temperature of extremities 
pain
outcome
Здесь выбросов нет

Принимаеи решение, что данные не содержат выбросов, только ошибку в столбце Age

In [103]:
pd.options.mode.chained_assignment = None

In [104]:
df.loc[df.age == 9, 'age'] = 2

In [105]:
df['age'].value_counts()

1    256
2     22
Name: age, dtype: int64

### surgery?

Может принимать только значение 1 или 2. Заменим на значение моды

In [106]:
df['surgery?'].fillna(df['surgery?'].mode()[0], inplace=True)

### temperature of extremities, pain
это субьективные оценки и сложно установить взаимосвять с другими показателями
### outcome
так же может принимать разные значения при равных условиях, поэтому заполним пропуски последующими значениями df

In [107]:
df.pain.fillna(method='ffill', inplace=True)
df['temperature of extremities'].fillna(method='ffill', inplace=True)
df.outcome.fillna(method='ffill', inplace=True)

### rectal temperature, respiratory rate, pulse

чтобы рассчитать данные показатели, добавим еще столбцы с округленными значениями до целого для температуры и до десятков для частоты дыхания и пульса. Таким образом мы сможем сгруппировать значения. 

In [108]:
df['pulse_round'] = (df['pulse']/10).round()*10
df['respiratory_rate_round'] = (df['respiratory rate']/10).round()*10
df['rectal_temperature_round'] = (df['rectal temperature']).round()

In [109]:
# частота дыхания повышается с учащением пульса
df.groupby('pulse_round')['respiratory rate'].median()

pulse_round
30.0     15.0
40.0     19.0
50.0     20.0
60.0     30.0
70.0     24.0
80.0     32.0
90.0     28.0
100.0    35.0
110.0    35.0
120.0    48.0
130.0    42.0
140.0    48.0
150.0    72.0
160.0    52.0
180.0    84.0
Name: respiratory rate, dtype: float64

In [110]:
# пульс учащается с отклонением от нормы
df.groupby('rectal_temperature_round')['pulse'].median()

rectal_temperature_round
35.0    140.0
36.0     88.0
37.0     60.0
38.0     60.0
39.0     66.0
40.0     92.0
41.0     72.0
Name: pulse, dtype: float64

In [111]:
# температура повышается и дыхание учащается
df.groupby('respiratory_rate_round')['rectal temperature'].mean()

respiratory_rate_round
10.0     37.914815
20.0     38.056790
30.0     38.056410
40.0     38.290909
50.0     38.510000
60.0     38.780000
70.0     38.580000
80.0     38.320000
90.0     38.933333
100.0    38.750000
Name: rectal temperature, dtype: float64

In [112]:
df['respiratory rate'].fillna(df.groupby('pulse_round')['respiratory rate'].transform('median'), inplace=True)
df['pulse'].fillna(df.groupby('rectal_temperature_round')['pulse'].transform('median'), inplace=True)
df['rectal temperature'].fillna(df.groupby('respiratory_rate_round')['rectal temperature'].transform('mean').round(1), inplace=True)

In [113]:
pd.concat([df.describe(), df.mode()])

surgery?         age  rectal temperature       pulse  \
count  278.000000  278.000000          268.000000  275.000000   
mean     1.402878    1.079137           38.170149   71.974545   
std      0.491361    0.270439            0.694551   28.548897   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.100000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    2.000000           40.800000  184.000000   
0        1.000000    1.000000           38.100000   48.000000   

       respiratory rate  temperature of extremities        pain     outcome  \
count        275.000000                  278.000000  278.000000  278.000000   
mean          30.363636                    2.356115    2.899281    1.550360   
std           16.780859                    1.047163    1.293108    0.742702   
min            8.000000                    1.000000    1.000000    1.000000   
25%           20.000000                    1.000000    2.000000    1.000000   
50%           28.000000                    3.000000    3.000000    1.000000   
75%           36.000000                    3.000000    4.000000    2.000000   
max           96.000000                    4.000000    5.000000    3.000000   
0             20.000000                    3.000000    3.000000    1.000000   

       pulse_round  respiratory_rate_round  rectal_temperature_round  
count   269.000000              240.000000                238.000000  
mean     71.933086               30.500000                 38.138655  
std      28.948925               18.104763                  0.780631  
min      30.000000               10.000000                 35.000000  
25%      50.000000               20.000000                 38.000000  
50%      60.000000               20.000000                 38.000000  
75%      90.000000               40.000000                 38.000000  
max     180.000000              100.000000                 41.000000  
0        50.000000               20.000000                 38.000000

In [117]:
# все еще есть пропуски, но их уже немного. Заменим на значение медианы

df['respiratory rate'].fillna(df['respiratory rate'].median(), inplace=True)
df['pulse'].fillna(df['pulse'].median(), inplace=True)
df['rectal temperature'].fillna(df['rectal temperature'].median(), inplace=True)

In [118]:
pd.concat([df.describe(), df.mode()])

surgery?         age  rectal temperature       pulse  \
count  278.000000  278.000000          278.000000  278.000000   
mean     1.402878    1.079137           38.167626   71.888489   
std      0.491361    0.270439            0.682025   28.405874   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.100000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    2.000000           40.800000  184.000000   
0        1.000000    1.000000           38.100000   48.000000   

       respiratory rate  temperature of extremities        pain     outcome  \
count        278.000000                  278.000000  278.000000  278.000000   
mean          30.338129                    2.356115    2.899281    1.550360   
std           16.691533                    1.047163    1.293108    0.742702   
min            8.000000                    1.000000    1.000000    1.000000   
25%           20.000000                    1.000000    2.000000    1.000000   
50%           28.000000                    3.000000    3.000000    1.000000   
75%           36.000000                    3.000000    4.000000    2.000000   
max           96.000000                    4.000000    5.000000    3.000000   
0             20.000000                    3.000000    3.000000    1.000000   

       pulse_round  respiratory_rate_round  rectal_temperature_round  
count   269.000000              240.000000                238.000000  
mean     71.933086               30.500000                 38.138655  
std      28.948925               18.104763                  0.780631  
min      30.000000               10.000000                 35.000000  
25%      50.000000               20.000000                 38.000000  
50%      60.000000               20.000000                 38.000000  
75%      90.000000               40.000000                 38.000000  
max     180.000000              100.000000                 41.000000  
0        50.000000               20.000000                 38.000000

In [99]:
pd.concat([df_start.describe(), df_start.mode()])

surgery?         age  rectal temperature       pulse  \
count  299.000000  300.000000          240.000000  276.000000   
mean     1.397993    1.640000           38.167917   71.913043   
std      0.490305    2.173972            0.732289   28.630557   
min      1.000000    1.000000           35.400000   30.000000   
25%      1.000000    1.000000           37.800000   48.000000   
50%      1.000000    1.000000           38.200000   64.000000   
75%      2.000000    1.000000           38.500000   88.000000   
max      2.000000    9.000000           40.800000  184.000000   
0        1.000000    1.000000           38.000000   48.000000   

       respiratory rate  temperature of extremities       pain     outcome  
count        242.000000                  244.000000  245.00000  299.000000  
mean          30.417355                    2.348361    2.95102    1.551839  
std           17.642231                    1.045054    1.30794    0.737187  
min            8.000000                    1.000000    1.00000    1.000000  
25%           18.500000                    1.000000    2.00000    1.000000  
50%           24.500000                    3.000000    3.00000    1.000000  
75%           36.000000                    3.000000    4.00000    2.000000  
max           96.000000                    4.000000    5.00000    3.000000  
0             20.000000                    3.000000    3.00000    1.000000